text/0000-const-locals.md

```diff
@@ -0,0 +1,63 @@
+- Feature Name: const_locals
+- Start Date: 2018-01-11
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow `let` bindings in the body of constants and const fns. Additionally enable
+destructuring in `let` bindings and const fn arguments.
+
+# Motivation
+[motivation]: #motivation
+
+It makes writing const fns much more like writing regular functions and is
+not possible right now because the old constant evaluator was a constant folder
+that could only process expressions. With the miri const evaluator this feature
+exists but is still disallowed.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`let` bindings in constants and const fn work just like `let` bindings
+everywhere else. Historically these did not exist in constants and const fn
```

Comment:
`let` bindings "everywhere else" (normal `fn`s) introduce bindings which have move semantics when
`typeof(binding) /: Copy`. Does this hold in `const fn`?

Reply:
Yes. All code is compiled normally in const fn. The only difference is that the execution is on virtual hardware instead of real hardware. So if your code fails to compile outside of a const environment, it will also fail to compile within a const environment. The reverse does not hold, there's a lot of code that compiles right now that will never compile in a const environment.

Reply:
Great 👍. I'd include an explanation of this in the RFC text =)